### PR TITLE
RCAL-913 Resolve numpy 2 issues

### LINKED ISSUE #1413: Update elpp steps to numpy 2
_Issue [RCAL-913](https://jira.stsci.edu/browse/RCAL-913) was created on JIRA by [David Davis](https://jira.stsci.edu/secure/ViewProfile.jspa?name=ddavis):_

Modify the steps in the roman elpp to be compatable with numpy 2

### PR DESCRIPTION
**Pull request by WilliamJamieson**
Resolves [RCAL-913](https://jira.stsci.edu/browse/RCAL-913)

Closes #1413

This PR together with spacetelescope/stcal#305 (already merged) resolves all the issues from numpy 2 detected by the unit tests.

Note, I need to run the regression tests to confirm I caught all the issues.

## Tasks
- [ ] **request a review from someone specific**, to avoid making the maintainers review every PR
- [ ] add a build milestone, i.e. `24Q4_B15` (use the [latest build](https://github.com/spacetelescope/romancal/milestones) if not sure)
- [ ] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [ ] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types)
  - [ ] update or add relevant tests
  - [ ] update relevant docstrings and / or `docs/` page
  - [ ] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/romancal.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/romancal/wiki/How-to-resolve-JIRA-issues)

<details><summary>news fragment change types...</summary>

  - ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
  - ``changes/<PR#>.docs.rst``
  - ``changes/<PR#>.stpipe.rst``
  - ``changes/<PR#>.associations.rst``
  - ``changes/<PR#>.scripts.rst``
  - ``changes/<PR#>.mosaic_pipeline.rst``
  - ``changes/<PR#>.patch_match.rst``

  ## steps
  - ``changes/<PR#>.dq_init.rst``
  - ``changes/<PR#>.saturation.rst``
  - ``changes/<PR#>.refpix.rst``
  - ``changes/<PR#>.linearity.rst``
  - ``changes/<PR#>.dark_current.rst``
  - ``changes/<PR#>.jump_detection.rst``
  - ``changes/<PR#>.ramp_fitting.rst``
  - ``changes/<PR#>.assign_wcs.rst``
  - ``changes/<PR#>.flatfield.rst``
  - ``changes/<PR#>.photom.rst``
  - ``changes/<PR#>.flux.rst``
  - ``changes/<PR#>.source_detection.rst``
  - ``changes/<PR#>.tweakreg.rst``
  - ``changes/<PR#>.skymatch.rst``
  - ``changes/<PR#>.outlier_detection.rst``
  - ``changes/<PR#>.resample.rst``
  - ``changes/<PR#>.source_catalog.rst``
</details>

### CODE REVIEW

**Comment by schlafly** [comment]
That looks good.  Could you say a word or two about needing to change the cross products from 2-vectors to 3-vectors?  Did numpy 2 just remove 2D cross products?  Thanks!

**Comment by WilliamJamieson** [comment]
> That looks good. Could you say a word or two about needing to change the cross products from 2-vectors to 3-vectors? Did numpy 2 just remove 2D cross products? Thanks!

It is marked as deprecated, and generates ~192000 warnings. It is depreciated as part of the lead into numpy 2 being fully python array API compliant, see numpy/numpy#24818. Cross products themselves are not really defined for products of 2D vectors in general; hence, the removal. In reality, this is just a shorthand for computing the determinant of the matrix made with the two vector. This is what the update should really be doing.

File: romancal/skymatch/skymatch_step.py [hunk 1]
```diff
@@ -179,7 +179,9 @@ def _set_sky_background(self, sky_image, step_status):
 
         image.meta.background.method = str(self.skymethod)
         image.meta.background.subtracted = self.subtract
-        image.meta.background.level = sky
+        # In numpy 2, the dtypes are more carefully controlled, so to match the
+        # schema the data type needs to be re-cast to float64.
+        image.meta.background.level = sky.astype(np.float64)
```
Comment (WilliamJamieson) [inline romancal/skymatch/skymatch_step.py:184]: The tighter adherence to `dtypes` means that this number should have been a `float32` before, but that violates the schemas: https://github.com/spacetelescope/rad/blob/b61c261deb1e21e13401e480a9790c6a212a0861/src/rad/resources/schemas/sky_background-1.0.0.yaml#L17. Should those be updated?

**Comment by WilliamJamieson** [comment]
These changes do not effect the regression tests when run with numpy 1, but the regression tests have slightly different results when run with numpy 2.

**Comment by schlafly** [comment]
This looks good to me.  FWIW, I inspected the regtests and they don't vary by more than a part in 10^8 or so for the small values; the rare large reported fractional differences are due to zeros becoming 10^-8s.  I did not actually try to figure out those differences---they're a little weird in that they affect only the pipeline regtests---but they are scientifically unimportant.

**Review by schlafly — APPROVED** [review]

**Pull request merged**